File: cpp/tests/test_heap.cpp
// Testes do heap mínimo, do heap máximo e do heapsort.
#include <algorithm>
#include <functional>
#include <random>

#include "../heap/heap.hpp"
#include "teste.hpp"

std::vector<int> aleatorios(std::mt19937& sorteio, int maximo_tamanho) {
    std::vector<int> valores(sorteio() % (maximo_tamanho + 1));
    for (int& valor : valores) valor = sorteio() % 100;
    return valores;
}

TESTE(heap_minimo_devolve_em_ordem_crescente) {
    std::mt19937 sorteio(88);
    for (int rodada = 0; rodada < 100; rodada++) {
        auto valores = aleatorios(sorteio, 30);
        HeapMinimo<int> heap;
        for (int valor : valores) {
            heap.inserir(valor);
            VERIFICAR(heap.eh_valido());
        }
        std::vector<int> saida;
        while (!heap.vazio()) saida.push_back(heap.remover());
        std::sort(valores.begin(), valores.end());
        VERIFICAR_IGUAL(saida, valores);
    }
}

TESTE(heap_maximo_devolve_em_ordem_decrescente) {
    std::mt19937 sorteio(89);
    auto valores = aleatorios(sorteio, 40);
    HeapMaximo<int> heap(valores);  // construção com heapify
    VERIFICAR(heap.eh_valido());
    std::vector<int> saida;
    while (!heap.vazio()) saida.push_back(heap.remover());
    std::sort(valores.begin(), valores.end(), std::greater<int>());
    VERIFICAR_IGUAL(saida, valores);
}

TESTE(construir_com_heapify) {
    std::mt19937 sorteio(90);
    for (int rodada = 0; rodada < 100; rodada++) {
        auto valores = aleatorios(sorteio, 25);
        HeapMinimo<int> heap(valores);
        VERIFICAR(heap.eh_valido());
        VERIFICAR_IGUAL(heap.tamanho(), static_cast<int>(valores.size()));
        if (!valores.empty()) VERIFICAR_IGUAL(heap.topo(), *std::min_element(valores.begin(), valores.end()));
    }
}

TESTE(heap_vazio) {
    HeapMinimo<int> heap;
    VERIFICAR(heap.vazio());
    VERIFICAR_LANCA(heap.remover(), std::out_of_range);
    VERIFICAR_LANCA(heap.topo(), std::out_of_range);
}

TESTE(heapsort_ordena_no_proprio_vetor) {
    std::mt19937 sorteio(91);
    for (int rodada = 0; rodada < 200; rodada++) {
        auto valores = aleatorios(sorteio, 40);
        auto esperado = valores;
        std::sort(esperado.begin(), esperado.end());
        heapsort(valores);
        VERIFICAR_IGUAL(valores, esperado);
    }
}

int main() { return rodar_testes(); }

File: python/tests/test_heap.py
"""Testes do heap mínimo, do heap máximo e do heapsort.

Rode a partir da pasta `python/`:  python3 -m unittest
"""

import random
import unittest

from heap import HeapMaximo, HeapMinimo, heapsort


class TestHeap(unittest.TestCase):
    def test_heap_minimo_devolve_em_ordem_crescente(self):
        sorteio = random.Random(88)
        for _ in range(100):
            valores = [sorteio.randrange(100) for _ in range(sorteio.randrange(30))]
            heap = HeapMinimo()
            for valor in valores:
                heap.inserir(valor)
                self.assertTrue(heap.eh_valido())
            self.assertEqual([heap.remover() for _ in range(len(valores))], sorted(valores))

    def test_heap_maximo_devolve_em_ordem_decrescente(self):
        sorteio = random.Random(89)
        valores = [sorteio.randrange(100) for _ in range(40)]
        heap = HeapMaximo(valores)  # construção com heapify
        self.assertTrue(heap.eh_valido())
        self.assertEqual([heap.remover() for _ in range(len(valores))], sorted(valores, reverse=True))

    def test_construir_com_heapify(self):
        sorteio = random.Random(90)
        for _ in range(100):
            valores = [sorteio.randrange(50) for _ in range(sorteio.randrange(25))]
            heap = HeapMinimo(valores)
            self.assertTrue(heap.eh_valido())
            self.assertEqual(sorted(heap), sorted(valores))
            if valores:
                self.assertEqual(heap.topo(), min(valores))

    def test_heap_vazio(self):
        heap = HeapMinimo()
        self.assertTrue(heap.vazio())
        with self.assertRaises(IndexError):
            heap.remover()
        with self.assertRaises(IndexError):
            heap.topo()

    def test_heapsort(self):
        sorteio = random.Random(91)
        for _ in range(200):
            valores = [sorteio.randrange(100) for _ in range(sorteio.randrange(40))]
            copia = list(valores)
            self.assertEqual(heapsort(copia), sorted(valores))
            self.assertEqual(copia, sorted(valores))  # ordena no próprio vetor


if __name__ == "__main__":
    unittest.main()
